### Task 10.1
* Name:LI WAN
* Student Number:223718804
* E-mail:s223718804@deakin.edu.au
* Course:SIT720
---

In [2]:
#Q1 Create a MLP model with 10 hidden layers using  "data.csv" dataset and report performances using appropriate metrics. 
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.neural_network import MLPClassifier
from sklearn.metrics import accuracy_score, classification_report

local_path = r'E:\2-学习\1-Deakin\24-T1\SIT720\Task\data_10.csv'
df=pd.read_csv(local_path, comment='#')
df

t_0  t_1  t_2  t_3  t_4  t_5  t_6  t_7  t_8  t_9  ...  t_91  t_92  \
0      112  274  158  215  274  158  215  298   76  208  ...    71   297   
1       82  208  187  208  172  117  172  117  172  117  ...    81   240   
2       16  110  240  117  240  117  240  117  240  117  ...    65   112   
3       82  208  187  208  172  117  172  117  172  117  ...   208   302   
4       82  240  117  240  117  240  117  240  117  172  ...   209   260   
...    ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...   ...   ...   
43871   82  240  117  240  117  240  117  240  117  172  ...   141   260   
43872   82  240  117  240  117  240  117  240  117  172  ...   159   224   
43873   82  240  117  240  117  240  117  240  117  172  ...   260   141   
43874   82  240  117  240  117  240  117  240  117  172  ...   141   260   
43875  112  274  158  215  274  158  215  298   76  208  ...    71   297   

       t_93  t_94  t_95  t_96  t_97  t_98  t_99  malware  
0       135   171   215    35   208    56    71        1  
1       117    71   297   135   171   215    35        1  
2       123    65   112   123    65   113   112        1  
3       208   302   187   208   302   228   302        1  
4        40   209   260   141   260   141   260        1  
...     ...   ...   ...   ...   ...   ...   ...      ...  
43871   141   260   141   260   141   260   141        1  
43872    82   159   224    82   159   224    82        1  
43873   260   141   260   141   260   141   260        1  
43874   141   260   141   260   141   260   141        1  
43875   135   171   215    35   208    56    71        1  

[43876 rows x 101 columns]

The accuracy score reported is 0.9826, which means that the model correctly classified 98.26% of the test instances.
However, the classification report provides more insights into the model's performance. It shows that the model has a high precision of 0.98 for the positive class (malware = 1), which means that when the model predicts an instance as malware, it is correct 98% of the time. However, the recall for the negative class (malware = 0) is very low at 0.33, which means that the model misses many instances of non-malware and incorrectly classifies them as malware.
The F1-score, which is the harmonic mean of precision and recall, is low for the negative class (0.50) and high for the positive class (0.99). This imbalance is likely due to the class imbalance in the dataset, where there are significantly more instances of malware (8550) compared to non-malware (226).
Overall, while the model has a high accuracy, it performs poorly in correctly identifying non-malware instances. This issue could potentially be addressed by using techniques such as oversampling the minority class or adjusting the classification threshold.

In [3]:
#Split the dataset into features and target variable
X = df.drop(columns=['malware'])  # Assuming 'malware' is the name of the target variable
y = df['malware']

# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Create MLP model with 10 hidden layers
mlp = MLPClassifier(hidden_layer_sizes=(10,), max_iter=1000, random_state=42)

# Train the model
mlp.fit(X_train, y_train)

# Predictions on the test set
y_pred = mlp.predict(X_test)

# Report performance metrics
accuracy = accuracy_score(y_test, y_pred)
print("Accuracy:", accuracy)

print("Classification Report:")
print(classification_report(y_test, y_pred))

Accuracy: 0.9826800364630811
Classification Report:
              precision    recall  f1-score   support

           0       0.99      0.33      0.50       226
           1       0.98      1.00      0.99      8550

    accuracy                           0.98      8776
   macro avg       0.98      0.67      0.74      8776
weighted avg       0.98      0.98      0.98      8776



In [5]:
# Q2 Analyse impact of different activation function with adam solver on the model.
# Define a list of activation functions to test
activation_functions = ['identity', 'logistic', 'tanh', 'relu']

# Dictionary to store results
results = {}

# Loop through each activation function
for activation in activation_functions:
    # Create MLP model with Adam solver and current activation function
    mlp = MLPClassifier(hidden_layer_sizes=(10,), activation=activation, solver='adam', max_iter=1000, random_state=42)
    
    # Train the model
    mlp.fit(X_train, y_train)
    
    # Predictions on the test set
    y_pred = mlp.predict(X_test)
    
    # Calculate accuracy
    accuracy = accuracy_score(y_test, y_pred)
    
    # Store results
    results[activation] = accuracy

# Print results
print("Results:")
for activation, accuracy in results.items():
    print(f"Activation Function: {activation}, Accuracy: {accuracy}")

Results:
Activation Function: identity, Accuracy: 0.9814266180492252
Activation Function: logistic, Accuracy: 0.9742479489516864
Activation Function: tanh, Accuracy: 0.9742479489516864
Activation Function: relu, Accuracy: 0.9826800364630811


The results show that the activation function with the highest accuracy on the test set is the 'relu' (Rectified Linear Unit) activation function, achieving an accuracy of 0.9827 (98.27%).
The 'identity' activation function comes second with an accuracy of 0.9814 (98.14%), followed by 'logistic' and 'tanh' activation functions, both achieving an accuracy of 0.9742 (97.42%).
These results suggest that the 'relu' activation function performs the best for this particular dataset and model architecture. The 'identity' activation function also performs reasonably well, while the 'logistic' and 'tanh' activation functions perform slightly worse.

Q3: Explain your findings and report the best performance.
Based on the results from the two Questions provided, we can make the following observations and report the best performance:

1. **Dataset Characteristics**:
   - The dataset appears to be imbalanced, with a significantly higher number of instances belonging to the positive class (malware = 1) compared to the negative class (malware = 0). Specifically, there are 8550 instances of malware and only 226 instances of non-malware.

2. **Model Performance on the Original Setup**:
   - The Multi-Layer Perceptron (MLP) classifier with 10 hidden layers and the default settings achieved an overall accuracy of 0.9826 (98.26%) on the test set.
   - However, the classification report revealed that the model performed poorly in correctly identifying non-malware instances, with a recall of only 0.33 for the negative class.
   - This issue is likely due to the class imbalance in the dataset, where the model is biased towards the majority class (malware).

3. **Impact of Activation Functions**:
   - The second question evaluated the impact of different activation functions (identity, logistic, tanh, and relu) on the model's performance.
   - The 'relu' (Rectified Linear Unit) activation function achieved the highest accuracy of 0.9827 (98.27%) on the test set.
   - The 'identity' activation function also performed well, with an accuracy of 0.9814 (98.14%).
   - The 'logistic' and 'tanh' activation functions had slightly lower accuracies of 0.9742 (97.42%).  
   
4. **Best Performance**:
   - Based on the results, the best performance was achieved by the MLP classifier with 10 hidden layers, using the 'relu' activation function and the Adam solver.
   - This configuration yielded an overall accuracy of 0.9827 (98.27%) on the test set.

However, it's important to note that while the model achieved high overall accuracy, its performance in correctly identifying non-malware instances was poor due to the class imbalance. To improve the model's performance on the minority class, additional techniques such as oversampling, undersampling, or adjusting the classification threshold may be necessary.

Additionally, it's recommended to evaluate the model's performance using metrics that are more suitable for imbalanced datasets, such as precision, recall, F1-score, and area under the ROC curve (AUC-ROC), rather than relying solely on accuracy.